compiler-lib/src/symtab.rs

```diff
@@ -32,18 +41,20 @@ impl<'a, T> Scoped<'a, T> {
             .ok_or(CannotLeaveRootScopeError)
     }
 
-    pub fn define(&mut self, sym: &Symbol<'a>, val: T) -> Result<(), RedefinitionError> {
-        use std::collections::hash_map::Entry;
-        // TODO This allows shadowing
-        match self.current_scope().entry(*sym) {
-            Entry::Occupied(_) => return Err(RedefinitionError),
-            Entry::Vacant(e) => e.insert(val),
-        };
+    pub fn define(&mut self, sym: S, val: T) -> Result<(), RedefinitionError> {
+        if let Some(existing_definition) = self.visible_definition(sym) {
+            return Err(RedefinitionError);
+        }
+        let insert_res = self.current_scope().insert(sym, val);
+        debug_assert!(
+            insert_res.is_none(),
+            "visible_defintions broken, current_scope contained a value"
+        );
         Ok(())
     }
 
-    pub fn visible_definition(&self, sym: Symbol<'a>) -> Option<&T> {
-        // TODO O(1) access to visible definition
+    pub fn visible_definition(&self, sym: S) -> Option<&T> {
```

Comment:
Btw, why is this method called like that? I'd prefer `get_visible_definition` or simpler `lookup`.

Reply:
it is on TIP of this branch, which didn't compile

Reply:
rust convention is *not* using the "get_" prefix for getters and a "set_" prefix for setters.